Entity browser widget

Berdir opened the pull request:
This adds an entity browser widget that allows to select and upload files through an entity browser. Quite a few changes beside the widget itself to make it nicer.

This depends on https://github.com/drupal-media/entity_browser/pull/134, which will need to be merged first.
- The Widget class itself. Supports description field (will add alt/title for images and the show checkbox later). I replaced the selection display plugin with specific code that displays each file in a table, with preview, the fields and buttons to edit and delete, just like in the default class. Reordering is also supported.
- Adds a Thumbnail view mode and view display configurations and some changes to base fields to be able to hide everything except image/name. We don't have a good way to show something for videos/audios right now, though. But the view based display is quite flexible, in my project, I'm for example showing a preview image field for videos. We can also made video/audio formatters visible in there but not sure if that makes sense. Maybe we should have a formatter that displays nothing but the mime icon.
- Template and preprocess cleanup, throwing out a lot of cruft from nodes and 7.x, so that the thumbnail preview looks clean.

![selection_016](https://cloud.githubusercontent.com/assets/40826/12115461/d61d371e-b3b2-11e5-9e6f-c68636ea909f.png)

slashrsm commented on src/Plugin/Field/FieldWidget/FileEntityBrowser.php:
```diff
@@ -0,0 +1,212 @@
+
+    $element['field_widget_display']['#access'] = FALSE;
+    $element['field_widget_display_settings']['#access'] = FALSE;
+
```
Extra newline.

Berdir replied:
Fixed now.

slashrsm commented on src/Plugin/Field/FieldWidget/FileEntityBrowser.php:
```diff
@@ -0,0 +1,212 @@
+      '#title' => t('File view mode'),
+      '#type' => 'select',
+      '#default_value' => $this->getSetting('view_mode'),
+      '#options' => \Drupal::service('entity_display.repository')->getViewModeOptions('file'),
```
Should we inject this?

Berdir replied:
We should. The problem is that overriding create + __construct means that any change that entity_browser makes there breaks this. For example, switching to injecting entity type manager would change the arguments and break the subclass. But that's partially just an excuse for being lazy :)

slashrsm replied:
Lol :)

slashrsm replied:
I get it. OK, let's leave it like this. We can change this in the future if needed. EB should be more stable at that point.

slashrsm commented on src/Plugin/Field/FieldWidget/FileEntityBrowser.php:
```diff
@@ -0,0 +1,212 @@
+    $view_mode = $this->getSetting('view_mode');
+
+    $view_builder = \Drupal::entityTypeManager()->getViewBuilder('file');
+    $config = \Drupal::config('file.settings');
```
Same as above?

slashrsm commented on src/Plugin/Field/FieldWidget/FileEntityBrowser.php:
```diff
@@ -0,0 +1,212 @@
+          'data-entity-id' => $entity->id()
+        ],
+        'display' => $display,
+        'meta' => [
```
Widget states it supports image fields too. We should probably add title and alt in that case?

Berdir replied:
Added that now, will push soon. Looks like this, with both enabled and required:

![selection_018](https://cloud.githubusercontent.com/assets/40826/12558769/b7f4dc12-c391-11e5-82e2-ae0ed1894819.png)

Also respecting the settings for showing/requiring those 3 fields.

slashrsm commented:
Didn't test it but went through the code. It looks great. I am wondering if it would make sense to include this table-like widget in EB itself and make it in a way that would allow us to easily add metadata column fields in other classes.

slashrsm commented:
Oh... and https://github.com/drupal-media/entity_browser/pull/134 is merged now.

Berdir committed:
Issue #2656400 by Berdir: Add entity browser widget for file and image fields

Berdir committed:
Remove whitespaces

Berdir commented on src/Plugin/Field/FieldWidget/FileEntityBrowser.php:
```diff
@@ -0,0 +1,254 @@
+   * {@inheritdoc}
+   */
+  function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+    $this->items = $items;
```
I thought I provided an update to add $items also to the new method but apparently I didn't. I can provide an update for that or we can keep it here, shouldn't really a problem and there could always be someone that needs something in there that's not provided.

Berdir commented:
Not sure how exactly you want to do this. I rebased and merged the original commit with a commit message that references the issue that I created: https://www.drupal.org/node/2656400

But I also added a new comment for some follow-up changes, that makes it easier to review. Then you can easily squash the commits together.

slashrsm commented:
I rebased, squashed and merged. This is the actual commit: https://github.com/drupal-media/file_entity/commit/af4131334c88a7e7bd045874070c79b9af7d842d

Thank you! This is great!
